pages/blog/2020-04-06-flash-of-unstyled-dark-theme.md

```diff
@@ -0,0 +1,162 @@
+
+```css
+:root {
+  --color-fg: #000;
+  --color-bg: #fff;
+}
+@media (prefers-color-scheme: dark) {
+  :root {
+    --color-fg: #fff;
+    --color-bg: #000;
+  }
+}
+```
+
+In your React components simply refer to your global custom properties.
+
+```jsx
+// Pseudo code
+const Text = ({ children }) {
+  <P css={{ color: "var(--color-fg" }}>
+    {children}
+  </P>
+}
+```
+
+Or if you're using good old global CSS, just do something like:
+
+```jsx
+body {
+  color: var(--color-fg);
+  background: var(--color-bg);
+}
+```
+
+As we've now moved the theming logic to CSS, there should be no flash as client-side rendering kicks in.
+
+One drawback is that CSS custom properties is not supported by Internet Explorer, but it can be [polyfilled](https://jhildenbiddle.github.io/css-vars-ponyfill/).
+
+
+## Adding more features
+
+The global CSS custom properties solution works well for the simple use case of rendering a single theme. What about the more complex use cases?
+
+### Alloa users to switch themes
+
+A common feature is to allow users to switch between themes after the initial rendering. For this kind of interactivity we need JavaScript. That does not imply that we need React, we simply need an interface to modify the global CSS custom properties. Let's start with extracting our theme into a JSON structure:
+
+```javascript
+var themes = {
+  light: {
+    "color-fg": "#000",
+    "color-bg": "#fff",
+  },
+  dark: {
+    "color-fg": "#000",
+    "color-bg": "#fff",
+  }
+};
+```
+
+Now you can add this piece of *render-blocking* JavaScript in your `<head>`
+
+```javascript
+// Get root <html> node style interface
+var root = document.documentElement.style;
+var darkMatcher = window.matchMedia("(prefers-color-scheme: dark)");
+var prefersColorScheme = darkMatcher.matches ? "dark" : "light";
+
+window.__onSetTheme = function() {};
+window.__setTheme = function(themeName) {
+  var theme = themes[themeName];
+  Object.keys(theme).forEach(function(key) {
+    // Set global custom properties on root element
+    root.setProperty("--" + key, theme[key]);
+    window.__onSetTheme(theme)
+  });
+}
+window.__setTheme(prefersColorScheme);
+```
+
+This will set the color scheme for the initial render and also expose a global `__setTheme` function that you can call from within your React UI. You can also override the `__onSetTheme` to subscribe to when a theme change is applied.
+
+
+### Persist user theme preference
+
+The next thing you might want to do is to persist the theme that the user has opted for. A good candidate for this is to use `localStorage`. Update the `__setTheme` function
+
+```javascript
+window.__setTheme(themeName) {
+  window.localStorage.setItem("themeName", themeName);
+  ...
+}
+```
+
+Update the `setTheme` function call to attempt to use the persisted theme first, otherwise pass the `prefersColorScheme`
+
+```javascript
+var savedTheme = window.localStorage.getItem('themeName');
```

Comment:
same here ?
I understatnd is just for illustrations but some comment should be done or some warning for when localstorage is not available.

Reply:
Good point